### PR TITLE
Add 'sbt scalastyleCheck' and use it in Jenkinsfile.

### PR DESCRIPTION
**Pull request by er1c**

### CODE REVIEW

File: build.sbt
Comment (gzm0) [inline build.sbt:83]: We already have machinery for this in `Build.scala`. Look for `allProjects`. Adding `scalastyle` should be a single line change.

Comment (er1c) [reply build.sbt:83]: @gzm0 thanks for the pointer.   However, I'm struggling with the correct SBT configuration since `scalastyle` isn't a `TaskKey` (https://github.com/scalastyle/scalastyle-sbt-plugin/blob/master/src/main/scala/org/scalastyle/sbt/Plugin.scala#L51)  

```
import org.scalastyle.sbt.ScalastylePlugin.autoImport._
val scalaStyleTask = taskKey[Unit]("scalaStyle")
scalaStyleTask := scalastyle.toTask("").value,

...

        val keys = Seq[TaskKey[_]](
            clean, headerCreate in Compile, headerCreate in Test,
            headerCheck in Compile, headerCheck in Test, scalaStyleTask in Test
        )

```

Results iin  something like:

```
[error]
[error]   scalaTestSuite2_12 / Test / scalaStyleTask from Test / scalaStyleTask (Build.scala:587)
[error]
[error]   scalaTestSuite2_13 / Test / scalaStyleTask from Test / scalaStyleTask (Build.scala:587)
[error]
[error]   scalastyle from scalaStyleTask ((build.Build.root) Build.scala:559)
[error]      Did you mean Compile / scalastyle ?

```

Any ideas?

Comment (gzm0) [reply build.sbt:83]: I didn't realize `scalaStyle` is an `InputTask`. If you want to define a `Task` like this, you need to put the task definition into `commonSettings` so it gets applied to all projects (n.b. you need to do this in all configurations).

Comment (er1c) [reply build.sbt:83]: @gzm0 I tried this, but couldn't get it working - spent several hours trying  to without luck. In the example code above, I have the `scalaStyleTask := scalastyle.toTask("").value,` in the command settings, or do you mean the actual val?

Comment (er1c) [reply build.sbt:83]: Is there any reason not to include the  scalastyle in the normal compile (e.g. https://www.scala-sbt.org/1.x/docs/Howto-Dynamic-Task.html) I guess I was just trying to add it as a separate task to avoid making any "compile" changes, but this seems more straightforward

Comment (sjrd) [reply build.sbt:83]: Yes, there is a reason: because we don't want to spend time running the Scalastyle checker every time we compile ;)

Comment (er1c) [reply build.sbt]: @sjrd that's fair, and also validates my intuition on not going down that path.  I *think* just figured it out, PR is updated, I'm doing some additional testing, is it worth updating the travis build to use this instead of the manual list?

File: project/Build.scala
Comment (gzm0) [inline project/Build.scala:255]: Does this really need to be `dyn`? Ugh. (I have never used `toTask` on an `InputTask`, so I'm genuinely asking).

File: project/Build.scala
Comment (gzm0) [inline project/Build.scala:256]: Does running this in the `Test` configuration also check the normal code? (Our `Jenkinsfile` suggests otherwise, but it might be over-checking).

Comment (gzm0) [reply project/Build.scala:256]: Also, write this:

```scala
(scalastyle in Test).toTask("")
```

And import `scalastyle` for the entire file?

File: DEVELOPING.md
Comment (gzm0) [inline DEVELOPING.md:106]: SJS doesn't do double spaces. Please be consistent with the rest of the file.

**Review by gzm0 — COMMENTED** [review]
> I'm doing some additional testing, is it worth updating the travis build to use this instead of the manual list?

Definitely, one less place we can forget to add projects :)

File: project/Build.scala
Comment (er1c) [reply project/Build.scala:256]: ahh yeah, I was originally trying to keep the name of the task `scalastyle` so was trying to avoid the double naming, but that didn't work out, I'll update the import appropriately.  I'll double check the `taskDyn` that was cargo cult on my part from examples I could find.

File: DEVELOPING.md
Comment (er1c) [reply DEVELOPING.md:106]: 👍

File: project/Build.scala
Comment (er1c) [reply project/Build.scala:256]: `taskDyn` is required here, nothing runs if `Def.task`

Comment (er1c) [reply project/Build.scala:256]: It looks like both in Test and in Compile are required :/ trying to figure out if I need to have two tasks or if I can put them both in the same one

Comment (gzm0) [reply project/Build.scala:256]: > nothing runs if Def.task

Nothing runs or it fails to compile? If it doesn't run, you might be creating a dangling value. Try:

```scala
scalaStyleCheck := (scalastyle in Test).toTask("")

// or

scalaStyleCheck := Def.task {
  (scalastyle in Test).toTask("").value
}.value
```

Comment (er1c) [reply project/Build.scala]: Adding the extra `.value` seemed to do the trick.

Full code got updated to:

```
      scalastyleCheck := Def.task {
        val _ = (scalastyle in Compile).toTask("").value
        (scalastyle in Test).toTask("").value
      }.value,
```

**Review by gzm0 — APPROVED** [review]

**Comment by er1c** [comment]
@gzm0 @sjrd I don't think the jenkinsbot picked up the `Jenkinsfile` changes, based upon my local testing I think https://github.com/scala-js/scala-js/pull/4172 will need to be merged before the `Jenkinsfile` commit gets merged.  Should I remove that commit from this PR to keep the `Build.scala` changes separate?

**Comment by sjrd** [comment]
No, indeed. The Jenkins bot doesn't "trust" non-maintainers to edit Jenkinsfile, so during PR validation it ignores the changes and keeps using the `Jenkinsfile` on master. I did not find a way to configure it otherwise.

**Comment by er1c** [comment]
I removed the `Jenkinsfile` changes to simplify this and still allow running the command locally for development.

**Comment by gzm0** [comment]
@sjrd I believe this is waiting on you.

**Review by sjrd — CHANGES_REQUESTED** [review]
Running `scalastyleCheck` fails on the following projects:
* library and libraryAux, but that would be fixed with a rebase I believe,
* scalalib and partest, for which it shouldn't be run.

The command should somehow do nothing for `scalalib` and `partest`. Otherwise running it is just confusing as it reports errors even though a PR would accept the code.

File: project/Build.scala
Comment (gzm0) [inline project/Build.scala:573]: Instead of this, consider simply overriding `scalastyleCheck` in the projects where we do not want it.

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: project/Build.scala
Comment (er1c) [reply project/Build.scala]: Updated, see if this approach matches what you are suggesting here

File: project/Build.scala
Comment (er1c) [inline project/Build.scala:2026]: extra newlines ugh

File: build.sbt
Comment (gzm0) [inline build.sbt:40]: I'm not sure this indirection is necessary. Invoking `scalastyleCheck` is just as fine, no?

Comment (er1c) [reply build.sbt:40]: `scalastyle` is the usual command people run, so it's a bit more "friendly", but no, it is not necessary, thoughts?  I originally had just `scalastyleCheck` but after finding that this worked, I added it for my own muscle memory.  I'll defer to you and @sjrd.  Should I remove it?

Comment (sjrd) [reply build.sbt:40]: I have a slight preference for not having the command alias, and running `scalastyleCheck` instead. This is mostly because I don't like command aliases that shadow tasks that otherwise exist.

Comment (er1c) [reply build.sbt]: Removed, and updated Jenkinsfile/etc to use `scalastyleCheck`.  Thanks!

**Comment by er1c** [comment]
@sjrd @gzm0 CI passing

**Review by gzm0 — APPROVED** [review]

**Review by sjrd — APPROVED** [review]

**Pull request merged**
